core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -289,13 +290,23 @@ private[spark] class ExecutorAllocationManager(
       s" tasksperexecutor: $tasksPerExecutor")
     val maxNeeded = math.ceil(numRunningOrPendingTasks * executorAllocationRatio /
       tasksPerExecutor).toInt
-    if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
+    val totalNeed = if (tasksPerExecutor > 1 && maxNeeded == 1 && pendingSpeculative > 0) {
```

Comment:
I think `totalNeed` is a little confusing. Maybe `maxNeededWithSpeculationLocalityOffset` or something else more descriptive like that?